modules/margin-protocol/src/lib.rs

```diff
@@ -178,6 +203,289 @@ impl<T: Trait> Module<T> {
 	}
 }
 
+type PriceResult = result::Result<Price, DispatchError>;
+type Fixed128Result = result::Result<Fixed128, DispatchError>;
+type BalanceResult = result::Result<Balance, DispatchError>;
+
+// Price helpers
+impl<T: Trait> Module<T> {
+	/// The price from oracle.
+	fn _price(base: CurrencyId, quote: CurrencyId) -> PriceResult {
+		T::PriceProvider::get_price(base, quote).ok_or(Error::<T>::NoPrice.into())
+	}
+
+	/// ask_price = price * (1 + ask_spread)
+	fn _ask_price(pool: LiquidityPoolId, held: CurrencyId, debit: CurrencyId, max: Option<Price>) -> PriceResult {
+		let price = Self::_price(debit, held)?;
+		//FIXME: liquidity pools should provide spread based on trading pair
+		let spread: Price = T::LiquidityPools::get_ask_spread(pool, held)
+			.ok_or(Error::<T>::NoAskSpread)?
+			.into();
+		let ask_price: Price = Price::from_natural(1).saturating_add(spread).saturating_mul(price);
+
+		if let Some(m) = max {
+			if ask_price > m {
+				return Err(Error::<T>::MarketPriceTooHigh.into());
+			}
+		}
+
+		Ok(ask_price)
+	}
+
+	/// bid_price = price * (1 - bid_spread)
+	fn _bid_price(pool: LiquidityPoolId, held: CurrencyId, debit: CurrencyId) -> PriceResult {
+		let price = Self::_price(debit, held)?;
+		//FIXME: liquidity pools should provide spread based on trading pair
+		let spread: Price = T::LiquidityPools::get_bid_spread(pool, held)
+			.ok_or(Error::<T>::NoAskSpread)?
+			.into();
+
+		Ok(Price::from_natural(1).saturating_sub(spread).saturating_mul(price))
+	}
+
+	/// usd_value = amount * price
+	fn _usd_value(currency_id: CurrencyId, amount: Fixed128) -> Fixed128Result {
+		let price = {
+			let p = Self::_price(CurrencyId::AUSD, currency_id)?;
+			fixed_128_from_fixed_u128(p)
+		};
+		amount.checked_mul(&price).ok_or(Error::<T>::NumOutOfBound.into())
+	}
+}
+
+// Trader helpers
+impl<T: Trait> Module<T> {
+	/// Unrealized profit and loss of a position(USD value).
+	///
+	/// unrealized_pl_of_position = (curr_price - open_price) * leveraged_held
+	fn _unrealized_pl_of_position(position: &Position<T>) -> Fixed128Result {
+		// open_price = abs(leveraged_debits / leveraged_held)
+		let open_price = position
+			.leveraged_debits
+			.checked_div(&position.leveraged_held)
+			.expect("ensured safe on open position")
+			.saturating_abs();
+		let curr_price = {
+			let p = Self::_bid_price(position.pool, position.pair.quote, position.pair.base)?;
```

Comment:
If `leverage.is_long()`,  should get `_bid_price`
if `leverage.is_short()`, should get `_ask_price`